src/Build/Instance/RunningObjectTable.cs

```diff
@@ -14,34 +16,29 @@ namespace Microsoft.Build.Execution
     /// <remarks>
     /// See https://docs.microsoft.com/en-us/windows/desktop/api/objidl/nn-objidl-irunningobjecttable.
     /// </remarks>
-    internal class RunningObjectTable : IDisposable, IRunningObjectTableWrapper
+    internal class RunningObjectTable : IRunningObjectTableWrapper
     {
-        private readonly IRunningObjectTable rot;
-        private bool isDisposed = false;
+        private readonly Thread _initRotThread;
+        private IRunningObjectTable _rot;
 
         public RunningObjectTable()
         {
-            Ole32.GetRunningObjectTable(0, out this.rot);
-        }
-
-        public void Dispose()
-        {
-            if (this.isDisposed)
-            {
-                return;
-            }
-
-            Marshal.ReleaseComObject(this.rot);
-            this.isDisposed = true;
+            // To avoid deadlock, create ROT in a MTA thread. And the
+            // object will be MTA
+            _initRotThread =
+                new Thread(() => Ole32.GetRunningObjectTable(0, out _rot));
```

Comment:
Such a short-lived thread seems wasteful. And if an exception is thrown from it, it will crash the process. Why not use Task.Run for this, and store the `Task<IRunningObjectTable>` as a field instead of the `Thread`?
Also, check the current thread's apartment and do it yourself directly if it's already an MTA.